Makefile

```diff
@@ -38,153 +48,181 @@ help:  ## Display this help
 
-.PHONY: fmt
-fmt: ## Run gofmt and write changes to each file
-	gofmt -l -w $(CODE_DIRS)
-
-.PHONY: fmt-check
-fmt-check: ## Run gofmt and report an error if any changes are made
-	./hack/gofmt.sh
-
-.PHONY: docs
-docs: $(patsubst %.dot,%.png,$(wildcard docs/*.dot))
+## --------------------------------------
+## Build/Run Targets
+## --------------------------------------
 
-%.png: %.dot
-	dot -Tpng $< >$@
+.PHONY: build
+build: generate manifests manager tools ## Build everything
 
-.PHONY: e2e-local
-e2e-local:
-	operator-sdk test local ./test/e2e \
-		--namespace $(TEST_NAMESPACE) \
-		--up-local $(SETUP) \
-		$(DEBUG) --go-test-flags "$(GO_TEST_FLAGS)"
+.PHONY: manager
+manager: generate fmt vet ## Build manager binary
+	go build -ldflags $(LDFLAGS) -o bin/manager main.go
 
 .PHONY: run
-run: ## Run the operator outside of a cluster in development mode
-	operator-sdk run --local \
-		--go-ldflags=$(LDFLAGS) \
-		--watch-namespace=$(RUN_NAMESPACE) \
-		--operator-flags="-dev"
+run: generate fmt vet manifests ## Run against the configured Kubernetes cluster in ~/.kube/config
+	go run -ldflags $(LDFLAGS) ./main.go -namespace=$(RUN_NAMESPACE) -dev
 
 .PHONY: demo
-demo: ## Run the operator outside of a cluster using the demo driver
-	operator-sdk run --local \
-		--go-ldflags=$(LDFLAGS) \
-		--watch-namespace=$(RUN_NAMESPACE) \
-		--operator-flags="-dev -demo-mode"
+demo: generate fmt vet manifests ## Run in demo mode
+	go run -ldflags $(LDFLAGS) ./main.go -namespace=$(RUN_NAMESPACE) -dev -demo-mode
+
+.PHONY: install
+install: $(KUSTOMIZE) manifests ## Install CRDs into a cluster
+	$(KUSTOMIZE) build config/crd | kubectl apply -f -
+
+.PHONY: uninstall
+uninstall: $(KUSTOMIZE) manifests ## Uninstall CRDs from a cluster
+	kustomize build config/crd | kubectl delete -f -
+
+.PHONY: deploy
+deploy: $(KUSTOMIZE) manifests ## Deploy controller in the configured Kubernetes cluster in ~/.kube/config
+	cd config/manager && kustomize edit set image controller=${IMG}
+	kustomize build config/default | kubectl apply -f -
+
+.PHONY: manifests
+manifests: controller-gen ## Generate manifests e.g. CRD, RBAC etc.
+	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+
+.PHONY: generate
+generate: controller-gen ## Generate code
+	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
+
+.PHONY: $(KUSTOMIZE)
+$(KUSTOMIZE):
+	cd $(TOOLS_DIR); ./install_kustomize.sh
+
+# find or download controller-gen
+# download controller-gen if necessary
+.PHONY: controller-gen
+controller-gen:
+ifeq (, $(shell which controller-gen))
+	@{ \
+	set -e ;\
+	CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
+	cd $$CONTROLLER_GEN_TMP_DIR ;\
+	go mod init tmp ;\
+	go get sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION) ;\
+	rm -rf $$CONTROLLER_GEN_TMP_DIR ;\
+	}
+CONTROLLER_GEN=$(GOBIN)/controller-gen
+else
+CONTROLLER_GEN=$(shell which controller-gen)
+endif
```

Comment:
This does not ensure that we are running with a specific version of controller-gen. If the user has an outdated controller-gen installed on his machine, it might result in confusing behaviour. Should the executable be placed in a dedicated location, and referred as such, like Kustomize ?

Reply:
We discussed this today in the community meeting and agreed that I will make this change as a follow-up PR.